tls/tls_ticket.c

```diff
@@ -22,409 +22,646 @@
+		T_ERR_NL("TLS: can't init ticket secret for vhost '%s'",
+			 vhost_name);
+		goto err;
+	}
 
-	p += cert_len;
+	for (i = 0; i < 2; i++) {
+		unsigned char kn_hash[TTLS_TICKET_KEY_LEN];
+		TlsTicketKey *key = &tcfg->keys[i];
+		unsigned long ts;
+
+		rwlock_init(&key->lock);
+		/*
+		 * Make a unique name for the key: mix vhost_name and key number
+		 * and ticket_key_name_iv. We don't need any cryptography safe
+		 * values here. Just something pretty unique, but equal on all
+		 * Tempesta noes with the same configuration. Since vhost name
+		 * is not something really random, anyone can deduce the value
+		 * behind the hash. It's not a problem, we just want to check,
+		 * that we have issued the ticket.
+		 */
+		r = ttls_md_starts(&md_ctx);
+		r |= ttls_md_update(&md_ctx, ticket_key_name_iv,
+				    SLEN(ticket_key_name_iv));
+		r |= ttls_md_update(&md_ctx, (unsigned char *)&i, sizeof(i));
+		r |= ttls_md_update(&md_ctx, vhost_name, vn_len);
+		r |= ttls_md_finish(&md_ctx, kn_hash);
+		if (r) {
+			T_ERR_NL("TLS: can't init ticket key name");
+			goto err;
+		}
+		memcpy(key->name, kn_hash,
+		       min(sizeof(key->name), sizeof(kn_hash)));
+		/*
+		 * The configuration is just being created or updated,
+		 * create current key, and previous key to allow resuming
+		 * sessions with clients who got the session before
+		 * (re-)configuration or switched from existent Tempesta nodes
+		 * to this fresh new one.
+		 */
+		ts = ttls_ticket_get_time(tcfg->lifetime);
+		if (i)
+			ts -= tcfg->lifetime;
+		if ((r = __ttls_ticket_gen_key(key, ts, tcfg->secret))) {
+			T_ERR_NL("TLS: can't init ticket key value");
+			goto err;
+		}
+	}
 
-	*olen = p - buf;
+	setup_timer(&tcfg->timer, ttls_ticket_rotate_keys, (unsigned long)tcfg);
+	secs = tcfg->lifetime - (tfw_current_timestamp() % tcfg->lifetime);
+	mod_timer(&tcfg->timer, jiffies + msecs_to_jiffies(secs * 1000));
 
-	return 0;
+err:
+	ttls_md_free(&md_ctx);
+	return r;
 }
 
-/*
- * Unserialise session, see ssl_save_session()
+/**
+ * Clean Session Ticket keys.
+ *
+ * Vhost is unloaded and to be deleted: safe to remove keys. Can be called in
+ * process context.
  */
-static int ssl_load_session(TlsSess *session,
-				 const unsigned char *buf, size_t len)
+int
+ttls_tickets_clean(TlsPeerCfg *cfg)
 {
-	const unsigned char *p = buf;
-	const unsigned char * const end = buf + len;
-	size_t cert_len;
+	TlsTicketPeerCfg *tcfg = &cfg->tickets;
 
-	if (p + sizeof(TlsSess) > end)
-		return(TTLS_ERR_BAD_INPUT_DATA);
+	del_timer_sync(&tcfg->timer);
+	/* Wipe the keys. */
+	memset(tcfg, 0, sizeof(TlsTicketPeerCfg));
 
-	memcpy(session, p, sizeof(TlsSess));
-	p += sizeof(TlsSess);
+	return 0;
+}
 
-	if (p + 3 > end)
-		return(TTLS_ERR_BAD_INPUT_DATA);
+/**
+ * Setup TLS tickets shared context.
+ */
+int
+ttls_tickets_init()
+{
+	int err = TTLS_ERR_BAD_INPUT_DATA;
 
-	cert_len = (p[0] << 16) | (p[1] << 8) | p[2];
-	p += 3;
+	t_cfg.cipher_info = ttls_cipher_info_from_type(TTLS_CIPHER_AES_256_GCM);
```

Comment:
I got ciphers recommendations from RFC 5077 in a wrong way. The ticket is short so AES128 is sufficient for the task, the RFC says the same.  _I will fix this in a separate PR_: although simple replace TTLS_CIPHER_AES_256_GCM -> TTLS_CIPHER_AES_128_GCM works, the crypto_aead_setkey() should call a hash function to shrink the 256b key to 128 bits. We definitely don't want that in hot path, so need to revise all structures and set appropriate key size.

Reply:
By the way, is there any chance to do something with SHA call in the perf top? Can we reduce number of call, e.g. by batching the calls? Also are the better implementations of SHA, e.g. in OpenSSL or WolfSSL?

Reply:
SHA is not highly used in the ticket hot path, it's used only during key rotation. But the abbreviated handshake in whole has less operations, while SHA checksum is still calculated for each handshake message, thus it's higher in `top` comparing to full handshake. 

I don't see how we can batch SHA calls. We call it for every incoming and outgoing handshake message, so we cant batch it for a single connection. It's possible to batch several connections, and calculate hashsums in parallel, but it confront our current architecture, when a every single packet is processed as soon at it  was received. This can be potential for GPU handshake processi